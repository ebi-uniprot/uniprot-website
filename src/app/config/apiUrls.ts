import joinUrl from 'url-join';

import { stringifyUrl } from '../../shared/utils/url';
import {
  getApiSortDirection,
  SortDirection,
} from '../../uniprotkb/types/resultsTypes';

import { apiPrefix } from '../../shared/config/apiUrls/apiPrefix';

// Release notes
export const entry = (id?: string) =>
  id && joinUrl(apiPrefix, 'release-notes', id);

export const search = ({
  query,
  sort,
}: Record<string, string[] | string | null>) =>
  stringifyUrl(joinUrl(apiPrefix, 'release-notes/search'), {
    query: [query || '*'].filter(Boolean).join(' AND '),
    sort: sort || `release_date ${getApiSortDirection(SortDirection.descend)}`,
  });

export default { releaseNotes: { entry, search } };
