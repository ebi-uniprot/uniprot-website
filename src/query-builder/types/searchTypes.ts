export enum Operator {
  AND = 'AND',
  OR = 'OR',
  NOT = 'NOT',
}

export enum ItemType {
  group = 'group',
  siblingGroup = 'sibling_group',
  single = 'single',
}

export enum DataType {
  boolean = 'boolean',
  date = 'date',
  enum = 'enum',
  integer = 'integer',
  string = 'string',
}

export enum FieldType {
  evidence = 'evidence',
  general = 'general',
  range = 'range',
}

type EvidenceGroup = {
  groupName: string;
  items: { name: string; code: string }[];
};

export type SearchTermType = {
  id: string;
  label: string;
  itemType: ItemType;
  term: string;
  dataType?: DataType;
  fieldType?: FieldType;
  example?: string;
  autoComplete?: string;
  autoCompleteQueryTerm?: string;
  regex?: string;
  evidenceGroups?: EvidenceGroup[];
  valuePrefix?: string;
  values?: {
    name: string;
    value: string;
  }[];
  items?: SearchTermType[];
  siblings?: SearchTermType[];
};

export type QueryBit = Record<string, string>;

export type Clause = {
  id: string;
  searchTerm: SearchTermType;
  queryBits: QueryBit;
  logicOperator: Operator;
};

export enum Evidence {
  GO = 'go',
  ANNOTATION = 'annotation',
}

export type EvidenceDataPoint = {
  groupName: string;
  items: { name: string; code: string }[];
};

export type Evidences = {
  [Evidence.GO]: EvidenceDataPoint[];
  [Evidence.ANNOTATION]: EvidenceDataPoint[];
};
