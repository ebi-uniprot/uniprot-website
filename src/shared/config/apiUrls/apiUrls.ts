import * as configure from './configure';
import * as entry from './entry';
import * as proteinsApi from './proteinsApi';
import * as releaseNotes from './releaseNotes';
import * as search from './search';
import * as results from './results';
import * as suggester from './suggester';

export default {
  configure,
  entry,
  proteinsApi,
  releaseNotes,
  results,
  search,
  suggester,
};
