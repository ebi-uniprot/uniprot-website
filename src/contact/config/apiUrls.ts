import joinUrl from 'url-join';

import { apiPrefix } from '../../shared/config/apiUrls/apiPrefix';

const token = joinUrl(apiPrefix, 'contact', 'token');

const send = joinUrl(apiPrefix, 'contact', 'send');

export default { token, send };
