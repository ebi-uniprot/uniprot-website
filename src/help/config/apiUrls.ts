import joinUrl from 'url-join';

import { stringifyUrl } from '../../shared/utils/url';

import { apiPrefix } from '../../shared/config/apiUrls/apiPrefix';
import { defaultFacets as helpDefaultFacets } from './HelpFacetConfiguration';
import { ApiDocsDefinition } from '../types/apiDocumentation';

const entry = (id?: string) => id && joinUrl(apiPrefix, 'help', id);

const search = ({
  query,
  sort,
  fields,
  queryFacets,
  facets = helpDefaultFacets,
  size,
}: Record<string, string[] | string | null>) =>
  stringifyUrl(joinUrl(apiPrefix, 'help/search'), {
    query: [
      query || '*',
      ...(Array.isArray(queryFacets)
        ? queryFacets
        : (queryFacets || '').split(',')
      )
        .filter(Boolean)
        // Sort in order to improve cache hits
        .sort()
        .map((facet) => {
          const [facetName, facetValue] = (facet || '').split(':');
          return `(${facetName}:${
            facetValue.includes(' ') ? `"${facetValue}"` : facetValue
          })`;
        }),
    ]
      .filter(Boolean)
      .join(' AND '),
    sort,
    fields,
    facets,
    size,
  });

const apiDocumentationDefinition = (d: ApiDocsDefinition) =>
  joinUrl(apiPrefix, d, 'api/docs');

export default { entry, search, apiDocumentationDefinition };
