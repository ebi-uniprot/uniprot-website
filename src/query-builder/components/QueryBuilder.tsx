import React, { FC, FormEvent, useState, useEffect } from 'react';
import {
  useHistory,
  useRouteMatch,
  generatePath,
  useLocation,
} from 'react-router-dom';
import { useDispatch } from 'react-redux';
import qs from 'query-string';
import { v1 } from 'uuid';
import { frame } from 'timing-functions';
import { PageIntro } from 'franklin-sites';

import ClauseList from './ClauseList';

import useDataApi from '../../shared/hooks/useDataApi';

import { createEmptyClause, defaultQueryFor } from '../utils/clause';
import { stringify } from '../utils/queryStringProcessor';
import parseAndMatchQuery from '../utils/parseAndMatchQuery';

import { addMessage } from '../../messages/state/messagesActions';

import apiUrls from '../../shared/config/apiUrls';
import { Namespace, NamespaceLabels } from '../../shared/types/namespaces';
import { LocationToPath, Location } from '../../app/config/urls';

import {
  MessageFormat,
  MessageLevel,
} from '../../messages/types/messagesTypes';
import { Clause, SearchTermType } from '../types/searchTypes';

import '../../uniprotkb/components/search/styles/search-container.scss';
import './styles/query-builder.scss';

const QueryBuilder: FC = () => {
  const history = useHistory();
  const location = useLocation();
  const match = useRouteMatch<{ namespace?: Namespace }>(
    LocationToPath[Location.QueryBuilder]
  );
  const dispatch = useDispatch();

  // To be replaced by getting it from url
  const [clauses, setClauses] = useState<Clause[]>([]);

  const namespace = match?.params?.namespace;

  const { loading, data: searchTermsData } = useDataApi<SearchTermType[]>(
    namespace && apiUrls.queryBuilderTerms(namespace)
  );

  useEffect(() => {
    setClauses([]);
  }, [namespace]);

  // if URL doesn't finish with a namespace redirect to "uniprotkb" by default
  useEffect(() => {
    if (!namespace) {
      history.replace(
        history.createHref({
          ...history.location,
          pathname: generatePath(LocationToPath[Location.QueryBuilder], {
            namespace: Namespace.uniprotkb,
          }),
        })
      );
    }
  }, [history, namespace]);

  useEffect(() => {
    if (!(searchTermsData && namespace) || loading) {
      return;
    }

    setClauses((clauses) => {
      if (clauses.length) {
        return clauses;
      }

      const query = qs.parse(location.search, { decode: true })?.query;

      const [validatedQuery, invalidClauses] = parseAndMatchQuery(
        query,
        searchTermsData
      );

      if (invalidClauses.length) {
        // wait for next frame because this was causing a React warning
        frame().then(() => {
          dispatch(
            addMessage({
              id: Array.isArray(query) ? query[0] : query ?? v1(),
              content: `Found ${invalidClauses.length} invalid query term${
                invalidClauses.length === 1 ? '' : 's'
              } for ${namespace}: ${invalidClauses
                .map((invalid) => `"${invalid.searchTerm.term}"`)
                .join(', ')}`,
              format: MessageFormat.POP_UP,
              level: MessageLevel.FAILURE,
            })
          );
        });
      }

      if (validatedQuery.length) {
        return validatedQuery;
      }

      return parseAndMatchQuery(defaultQueryFor(namespace), searchTermsData)[0];
    });
  }, [dispatch, location.search, loading, namespace, searchTermsData]);

  if (!searchTermsData || !namespace) {
    return null;
  }

  const addClause = () => {
    setClauses((clauses) => [...clauses, createEmptyClause()]);
  };

  const removeClause = (clauseId: string) => {
    setClauses((clauses) => {
      if (clauses.length === 1) {
        return [createEmptyClause()];
      }
      return clauses.filter((clause) => clause.id !== clauseId);
    });
  };

  const handleSubmit = (event: FormEvent) => {
    event.preventDefault();
    const queryString = stringify(clauses) || '*';
    history.push({
      pathname: `/${namespace}`,
      search: `query=${queryString}`,
    });
  };

  return (
    <>
      <PageIntro title="Advanced search" />

      <form
        className="query-builder"
        onSubmit={handleSubmit}
        data-testid="query-builder-form"
      >
        <fieldset>
          <label htmlFor="namespace-select">
            Searching in
            <select
              id="namespace-select"
              onChange={(e) => {
                history.replace(
                  history.createHref({
                    pathname: generatePath(
                      LocationToPath[Location.QueryBuilder],
                      { namespace: e.target.value }
                    ),
                  })
                );
              }}
              value={namespace}
            >
              {Object.keys(NamespaceLabels).map((key) => (
                <option value={key} key={key}>
                  {NamespaceLabels[key as Namespace]}
                </option>
              ))}
            </select>
          </label>
        </fieldset>
        <fieldset>
          <ClauseList
            removeClause={removeClause}
            clauses={clauses}
            setClauses={setClauses}
            searchTerms={searchTermsData}
          />
        </fieldset>
        <div className="query-builder__actions">
          <button
            type="button"
            id="add-field"
            className="button tertiary"
            data-testid="query-builder-add-field"
            onClick={addClause}
          >
            Add Field
          </button>
          <button type="submit" id="submit-query" className="button">
            Search
          </button>
        </div>
      </form>
    </>
  );
};

export default QueryBuilder;
