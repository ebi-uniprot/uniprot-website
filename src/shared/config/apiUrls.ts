import queryString from 'query-string';
import joinUrl from 'url-join';

import {
  getApiSortDirection,
  SortDirection,
  SelectedFacet,
  FileFormat,
  fileFormatsWithColumns,
  fileFormatToUrlParameter,
} from '../../uniprotkb/types/resultsTypes';
import { SortableColumn } from '../../uniprotkb/types/columnTypes';
import { BlastFacet } from '../../tools/blast/types/blastResults';
import { Namespace } from '../types/namespaces';

export const devPrefix = 'https://wwwdev.ebi.ac.uk';
export const prodPrefix = 'https://www.ebi.ac.uk';

const apiUrls = {
  // uniprotkb query builder terms
  queryBuilderTerms: (namespace: Namespace) =>
    joinUrl(devPrefix, `/uniprot/api/configure/${namespace}/search-fields`),
  // Annotation evidence used by query builder
  evidences: {
    annotation: joinUrl(
      devPrefix,
      '/uniprot/api/configure/uniprotkb/annotation_evidences'
    ),
    // Go evidences used by advanced go search
    // "itemType": "goterm",
    go: joinUrl(devPrefix, '/uniprot/api/configure/uniprotkb/go_evidences'),
  },
  // Database cross references used by query builder
  databaseXrefs: joinUrl(
    devPrefix,
    '/uniprot/api/configure/uniprotkb/databases'
  ),
  // Database cross reference fields in result column configure
  // "itemType": "database",
  databaseFields: joinUrl(
    devPrefix,
    '/uniprot/api/configure/uniprotkb/databasefields'
  ),
  // All result fields except database cross reference fields
  resultsFields: joinUrl(
    devPrefix,
    '/uniprot/api/configure/uniprotkb/result-fields'
  ),
  // Retrieve results
  search: (namespace: Namespace = Namespace.uniprotkb) =>
    joinUrl(devPrefix, `/uniprot/api/${namespace}/search`),
  download: joinUrl(devPrefix, '/uniprot/api/uniprotkb/stream'),
  variation: joinUrl(prodPrefix, '/proteins/api/variation'),
  features: joinUrl(prodPrefix, '/proteins/api/features'),
  accessions: joinUrl(devPrefix, '/uniprot/api/uniprotkb/accessions'),
  genecentric: (accession: string) =>
    joinUrl(devPrefix, '/uniprot/api/genecentric/', accession),

  entry: (accession: string) =>
    joinUrl(devPrefix, '/uniprot/api/uniprotkb/accession', accession),
  sequenceFasta: (accession: string) => `${apiUrls.entry(accession)}.fasta`,
  entryDownload: (accession: string, format: FileFormat) =>
    format === FileFormat.fastaCanonicalIsoform
      ? `${apiUrls.search()}?${queryString.stringify({
          query: `accession:${accession}`,
          includeIsoform: true,
          format: fileFormatToUrlParameter.get(
            FileFormat.fastaCanonicalIsoform
          ),
        })}`
      : `${apiUrls.entry(accession)}.${fileFormatToUrlParameter.get(format)}`,
  entryPublications: (accession: string) =>
    joinUrl(
      devPrefix,
      '/uniprot/api/uniprotkb/accession',
      accession,
      '/publications'
    ),
  organismSuggester: '/uniprot/api/suggester?dict=organism&query=?',

  // TODO: move that to UniRef-specific file?
  uniref: {
    entry: (id?: string) => id && joinUrl(devPrefix, '/uniprot/api/uniref', id),
  },
};

export default apiUrls;

const RE_QUERY = /\?$/;
export const getSuggesterUrl = (url: string, value: string) =>
  joinUrl(devPrefix, url.replace(RE_QUERY, value));

export const createFacetsQueryString = (facets: SelectedFacet[]) =>
  /**
   * Add double quotes to facet values which contain
   * spaces as otherwise the backend doesn't escape special characters
   * such as '.' or '-'.
   * Single word values shouldn't have double quotes as they can be boolean.
   * Range queries (/^\[.*]$/) should not have double quotes either.
   * */
  facets
    .map(
      (facet) =>
        `(${facet.name}:${
          facet.value.indexOf(' ') >= 0 && !facet.value.match(/^\[.*\]$/)
            ? `"${facet.value}"`
            : facet.value
        })`
    )
    .join(' AND ');

export const createAccessionsQueryString = (accessions: string[]) =>
  accessions
    .map((accession) => `accession:${accession}`)
    .sort() // to improve possible cache hit
    .join(' OR ');

const defaultFacets = new Map<Namespace, string[]>([
  [
    Namespace.uniprotkb,
    [
      'reviewed',
      'model_organism',
      'proteins_with',
      'existence',
      'annotation_score',
      'length',
    ],
  ],
  [Namespace.uniref, ['identity']],
]);
type QueryUrlProps = {
  namespace?: Namespace;
  query?: string;
  columns?: string[] | null;
  selectedFacets?: SelectedFacet[];
  sortColumn?: SortableColumn;
  sortDirection?: SortDirection;
  facets?: string[];
  size?: number;
};
export const getAPIQueryUrl = ({
  namespace = Namespace.uniprotkb,
  query = '*',
  columns = [],
  selectedFacets = [],
  sortColumn = undefined,
  sortDirection = SortDirection.ascend,
  facets,
  size,
}: QueryUrlProps = {}) => {
  let facetField = facets;
  if (!facetField) {
    facetField = defaultFacets.get(namespace);
  }
  return `${apiUrls.search(namespace)}?${queryString.stringify({
    size,
    query: `${[query || '*', createFacetsQueryString(selectedFacets)]
      .filter(Boolean)
      .join(' AND ')}`,
    fields: columns?.join(',') || undefined,
    facets: facetField?.join(','),
    sort:
      sortColumn &&
      `${sortColumn} ${getApiSortDirection(SortDirection[sortDirection])}`,
  })}`;
};

const localBlastFacets = Object.values(BlastFacet) as string[];
const excludeLocalBlastFacets = ({ name }: SelectedFacet) =>
  !localBlastFacets.includes(name);

type GetOptions = {
  columns?: string[];
  selectedFacets?: SelectedFacet[];
  sortColumn?: SortableColumn;
  sortDirection?: SortDirection;
  facets?: string[];
  size?: number;
};

export const getFeaturesURL = (accessions?: string[]) => {
  if (!(accessions && accessions.length)) {
    return null;
  }
  return `${apiUrls.features}?${queryString.stringify({
    // sort to improve possible cache hit
    accession: Array.from(accessions).sort().join(','),
  })}`;
};

export const getAccessionsURL = (
  accessions?: string[],
  {
    columns = [],
    selectedFacets = [],
    sortColumn = undefined,
    sortDirection = SortDirection.ascend,
    facets = defaultFacets.get(Namespace.uniprotkb),
    size,
  }: GetOptions = {}
) => {
  if (!(accessions && accessions.length)) {
    return null;
  }
  return `${apiUrls.accessions}?${queryString.stringify({
    size,
    // sort to improve possible cache hit
    accessions: Array.from(accessions).sort().join(','),
    facetFilter:
      createFacetsQueryString(selectedFacets.filter(excludeLocalBlastFacets)) ||
      undefined,
    fields: (columns && columns.join(',')) || undefined,
    facets: facets?.join(',') || undefined,
    sort:
      sortColumn &&
      `${sortColumn} ${getApiSortDirection(SortDirection[sortDirection])}`,
  })}`;
};

type GetUniProtPublicationsQueryUrl = {
  accession: string;
  selectedFacets: SelectedFacet[];
  size?: number;
};
export const getUniProtPublicationsQueryUrl = ({
  accession,
  selectedFacets,
  size,
}: GetUniProtPublicationsQueryUrl) =>
  `${apiUrls.entryPublications(accession)}?${queryString.stringify({
    facets: 'source,category,study_type',
    query:
      selectedFacets
        .map((facet) => `(${facet.name}:"${facet.value}")`)
        .join(' AND ') || undefined,
    size,
  })}`;

export const getDownloadUrl = ({
  query,
  columns,
  selectedFacets,
  sortColumn,
  sortDirection = SortDirection.ascend,
  fileFormat,
  compressed = false,
  size,
  selectedAccessions = [],
}: {
  query: string;
  columns: string[];
  selectedFacets: SelectedFacet[];
  sortColumn: SortableColumn;
  sortDirection: SortDirection;
  fileFormat: FileFormat;
  compressed: boolean;
  size?: number;
  selectedAccessions: string[];
}) => {
  const parameters: {
    query: string;
    format: string;
    fields?: string;
    sort?: string;
    includeIsoform?: boolean;
    size?: number;
    compressed?: boolean;
    download: true;
  } = {
    query: selectedAccessions.length
      ? createAccessionsQueryString(selectedAccessions)
      : `${query}${createFacetsQueryString(selectedFacets)}`,
    // fallback to json if something goes wrong
    format: fileFormatToUrlParameter.get(fileFormat) || 'json',
    download: true,
  };
  const isColumnFileFormat = fileFormatsWithColumns.includes(fileFormat);
  if (isColumnFileFormat && sortColumn) {
    parameters.sort = `${sortColumn} ${getApiSortDirection(
      SortDirection[sortDirection]
    )}`;
  }
  if (fileFormat === FileFormat.fastaCanonicalIsoform) {
    parameters.includeIsoform = true;
  }
  if (isColumnFileFormat && columns) {
    parameters.fields = columns.join(',');
  }
  if (size && !selectedAccessions.length) {
    parameters.size = size;
  }
  if (compressed) {
    parameters.compressed = true;
  }
  return `${apiUrls.download}?${queryString.stringify(parameters)}`;
};

export const literatureApiUrls = {
  literature: joinUrl(devPrefix, '/uniprot/api/literature'),
};

export const getPublicationURL = (id: string) =>
  joinUrl(literatureApiUrls.literature, id);

export const getPublicationsURL = (ids: string[]) =>
  `${literatureApiUrls.literature}/search?query=(${ids
    .map((id) => `id:${id}`)
    .join(' OR ')})`;

export const getProteinsApiUrl = (accession: string) =>
  `https://www.ebi.ac.uk/proteins/api/proteins/${accession}`;
