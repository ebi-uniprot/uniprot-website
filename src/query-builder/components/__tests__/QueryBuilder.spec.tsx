import React from 'react';
import { createMemoryHistory } from 'history';
import { act, fireEvent, screen, getByText } from '@testing-library/react';

import QueryBuilder from '../QueryBuilder';

import { resetUuidV1 } from '../../../../__mocks__/uuid';
import renderWithRedux from '../../../shared/__test-helpers__/RenderWithRedux';
import apiUrls from '../../../shared/config/apiUrls';
import { Namespace } from '../../../shared/types/namespaces';
import searchTermData from './__mocks__/configure_search-term.json';

import useDataApi from '../../../shared/hooks/useDataApi';
jest.mock('../../../shared/hooks/useDataApi');

// const mock = useDataApi;
// mock
//   .onGet(apiUrls.queryBuilderTerms(Namespace.uniprotkb))
//   .reply(200, searchTermData);

let rendered;

const history = createMemoryHistory();

describe('QueryBuilder', () => {
  beforeEach(async () => {
    resetUuidV1();
    useDataApi.mockReturnValue({
      data: searchTermData,
    });

    await act(async () => {
      rendered = renderWithRedux(<QueryBuilder />, { history });
    });
  });

  test('should render', () => {
    const { asFragment } = rendered;
    expect(asFragment()).toMatchSnapshot();
  });

  test('should add a clause', () => {
    const { getAllByTestId, getByTestId } = rendered;
    fireEvent.click(getByTestId('query-builder-add-field'));
    const clauses = getAllByTestId('search__clause');
    expect(clauses.length).toBe(5);
  });

  test('should remove a clause', () => {
    const { getAllByTestId } = rendered;
    fireEvent.click(getAllByTestId('clause-list-button-remove')[1]);
    const clauses = getAllByTestId('search__clause');
    expect(clauses.length).toBe(3);
  });

  test('should select field', () => {
    let entryNameField = screen.queryByPlaceholderText('P53_HUMAN');
    expect(entryNameField).toBeFalsy();
    const dropdownButton = screen.getByText(/All/, {
      selector: '.button.dropdown',
    });
    const clause = dropdownButton.closest('[data-testid="search__clause"]');
    fireEvent.click(dropdownButton);
    let entryNameFieldOption = getByText(clause, /Entry Name \[ID\]/);
    fireEvent.click(entryNameFieldOption);
    entryNameField = screen.queryByPlaceholderText('P53_HUMAN');
    expect(entryNameField).toBeTruthy();
  });

  test('should submit a simple query', () => {
    const { getByTestId, getAllByTestId, getByPlaceholderText } = rendered;
    const input = getByPlaceholderText(/ydj1/);
    fireEvent.change(input, { target: { value: 'zen' } });
    const all = getByPlaceholderText(/a4_human/);
    fireEvent.change(all, { target: { value: 'eve' } });
    const reviewedLogic = getAllByTestId('query-builder-logic-select');
    // Note this should be 1 as the first line shouldn't have one
    fireEvent.change(reviewedLogic[3], { target: { value: 'OR' } });
    // FIXME: These next lines are the ones triggering the warning even though
    // FIXME: we are indeed using act...
    act(() => {
      fireEvent.submit(getByTestId('query-builder-form'));
    });
    expect(history.location.search).toBe('?query=(gene:zen) OR eve');
  });
});
