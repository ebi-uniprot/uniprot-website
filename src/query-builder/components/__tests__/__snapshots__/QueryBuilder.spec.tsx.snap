// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryBuilder should render 1`] = `
<DocumentFragment>
  <div
    class="page-intro"
  >
    <h2>
      <button
        class="dropdown"
        type="button"
      >
        <svg
          fill="currentColor"
          height="20"
          viewBox="0 0 512 311"
          width="20"
        >
          <path
            d="M236 302L9 75a28 28 0 010-40L35 9C46-2 64-2 75 9l181 180L437 9c11-11 29-11 40 0l26 26c11 11 11 29 0 40L276 302a28 28 0 01-40 0z"
          />
        </svg>
         Advanced search
      </button>
    </h2>
    <div
      class="intro-content"
    >
      <div
        class="intro-links"
      />
    </div>
  </div>
  <form
    class="query-builder"
    data-testid="query-builder-form"
  >
    <fieldset>
      <label
        for="namespace-select"
      >
        Searching in
        <select
          id="namespace-select"
        >
          <option
            value="uniprotkb"
          >
            UniProtKB
          </option>
          <option
            value="uniref"
          >
            UniRef
          </option>
          <option
            value="uniparc"
          >
            UniParc
          </option>
          <option
            value="proteomes"
          >
            Proteomes
          </option>
          <option
            value="publications"
          >
            Publications
          </option>
          <option
            value="keywords"
          >
            Keywords
          </option>
        </select>
      </label>
    </fieldset>
    <fieldset>
      <div
        class="query-builder__clause"
        data-testid="search__clause"
      >
        <select
          class="query-builder__logic"
          data-testid="query-builder-logic-select"
        >
          <option
            value="AND"
          >
            AND
          </option>
          <option
            value="OR"
          >
            OR
          </option>
          <option
            value="NOT"
          >
            NOT
          </option>
        </select>
        <div
          class="dropdown-container"
        >
          <button
            class="button dropdown"
            type="button"
          >
            Gene Name [GN]
          </button>
          <div
            class="dropdown-menu"
          />
        </div>
        <div
          class="query-builder__inputs"
        >
          <label
            for="input_gene"
          >
            Gene Name [GN]
            <input
              id="input_gene"
              placeholder="ydj1"
              type="text"
              value=""
            />
          </label>
        </div>
        <button
          class="button tertiary button-remove"
          data-testid="clause-list-button-remove"
          type="button"
        >
          Remove
        </button>
      </div>
      <div
        class="query-builder__clause"
        data-testid="search__clause"
      >
        <select
          class="query-builder__logic"
          data-testid="query-builder-logic-select"
        >
          <option
            value="AND"
          >
            AND
          </option>
          <option
            value="OR"
          >
            OR
          </option>
          <option
            value="NOT"
          >
            NOT
          </option>
        </select>
        <div
          class="dropdown-container"
        >
          <button
            class="button dropdown"
            type="button"
          >
            Taxonomy [OC]
          </button>
          <div
            class="dropdown-menu"
          />
        </div>
        <div
          class="query-builder__inputs"
        >
          <label
            for="4"
          >
            Taxonomy [OC]
            <div
              class="autocomplete-container"
            >
              <span
                class="search-input"
              >
                <input
                  data-testid="search-input"
                  placeholder=""
                  type="text"
                  value=""
                />
                <span
                  class="search-input__suffix"
                  data-testid="search-input-suffix"
                  role="presentation"
                >
                  <svg
                    fill="currentColor"
                    height="14"
                    viewBox="0 0 512 512"
                    width="14"
                  >
                    <path
                      d="M505 443L405 343c-4-4-10-7-17-7h-16a208 208 0 10-36 36v16c0 7 3 13 7 17l100 100c9 9 24 9 34 0l28-28c9-10 9-25 0-34zM208 336a128 128 0 110-256 128 128 0 010 256z"
                    />
                  </svg>
                </span>
              </span>
              <div
                class="autocomplete-menu"
              >
                <div
                  class="dropdown-menu__panel"
                />
              </div>
            </div>
          </label>
        </div>
        <button
          class="button tertiary button-remove"
          data-testid="clause-list-button-remove"
          type="button"
        >
          Remove
        </button>
      </div>
      <div
        class="query-builder__clause"
        data-testid="search__clause"
      >
        <select
          class="query-builder__logic"
          data-testid="query-builder-logic-select"
        >
          <option
            value="AND"
          >
            AND
          </option>
          <option
            value="OR"
          >
            OR
          </option>
          <option
            value="NOT"
          >
            NOT
          </option>
        </select>
        <div
          class="dropdown-container"
        >
          <button
            class="button dropdown"
            type="button"
          >
            Keyword [KW]
          </button>
          <div
            class="dropdown-menu"
          />
        </div>
        <div
          class="query-builder__inputs"
        >
          <label
            for="5"
          >
            Keyword [KW]
            <div
              class="autocomplete-container"
            >
              <span
                class="search-input"
              >
                <input
                  data-testid="search-input"
                  placeholder=""
                  type="text"
                  value=""
                />
                <span
                  class="search-input__suffix"
                  data-testid="search-input-suffix"
                  role="presentation"
                >
                  <svg
                    fill="currentColor"
                    height="14"
                    viewBox="0 0 512 512"
                    width="14"
                  >
                    <path
                      d="M505 443L405 343c-4-4-10-7-17-7h-16a208 208 0 10-36 36v16c0 7 3 13 7 17l100 100c9 9 24 9 34 0l28-28c9-10 9-25 0-34zM208 336a128 128 0 110-256 128 128 0 010 256z"
                    />
                  </svg>
                </span>
              </span>
              <div
                class="autocomplete-menu"
              >
                <div
                  class="dropdown-menu__panel"
                />
              </div>
            </div>
          </label>
        </div>
        <button
          class="button tertiary button-remove"
          data-testid="clause-list-button-remove"
          type="button"
        >
          Remove
        </button>
      </div>
      <div
        class="query-builder__clause"
        data-testid="search__clause"
      >
        <select
          class="query-builder__logic"
          data-testid="query-builder-logic-select"
        >
          <option
            value="AND"
          >
            AND
          </option>
          <option
            value="OR"
          >
            OR
          </option>
          <option
            value="NOT"
          >
            NOT
          </option>
        </select>
        <div
          class="dropdown-container"
        >
          <button
            class="button dropdown"
            type="button"
          >
            All
          </button>
          <div
            class="dropdown-menu"
          />
        </div>
        <div
          class="query-builder__inputs"
        >
          <label
            for="input_All"
          >
            All
            <input
              id="input_All"
              placeholder="a4_human, P05067, cdc7 human"
              type="text"
              value=""
            />
          </label>
        </div>
        <button
          class="button tertiary button-remove"
          data-testid="clause-list-button-remove"
          type="button"
        >
          Remove
        </button>
      </div>
    </fieldset>
    <div
      class="query-builder__actions"
    >
      <button
        class="button tertiary"
        data-testid="query-builder-add-field"
        id="add-field"
        type="button"
      >
        Add Field
      </button>
      <button
        class="button"
        id="submit-query"
        type="submit"
      >
        Search
      </button>
    </div>
  </form>
</DocumentFragment>
`;
